src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -557,11 +610,31 @@
                                                extensions)]
     (write-cert-to-inventory! hostcert (:cert-inventory ca-settings))
     (utils/key->pem! public-key (:hostpubkey settings))
-    (utils/key->pem! private-key (:hostprivkey settings))
+    (utils/key->pem! private-key
+     (create-file-with-perms (:hostprivkey settings) p-key-perms))
     (utils/cert->pem! hostcert (:hostcert settings))
     (utils/cert->pem! hostcert
                       (path-to-cert (:signeddir ca-settings) certname))))
 
+(schema/defn ^:always-validate
+  ensure-master-file-perms!
+  [settings :- MasterSettings]
+  (let [p-key-dir (:privatekeydir settings)
+        cur-dir-perms (get-file-perms p-key-dir)]
+    (when-not (= p-key-dir-perms cur-dir-perms)
+      (set-file-perms p-key-dir p-key-dir-perms)
+      (log/warnf (str "The private key directory at '%s' had permissions set as "
+                      "'%s' and should be set to '%s' this has been corrected")
+                 p-key-dir cur-dir-perms p-key-dir-perms))
+    (doseq [key-file (fs/find-files p-key-dir #".*pem$")]
+      (let [key-path (.getPath key-file)
+            cur-key-perms (get-file-perms key-path)]
+        (when-not (= p-key-perms cur-key-perms)
+          (set-file-perms key-path p-key-perms)
+          (log/warnf (str "The private key located at '%s' had permissions set as "
+                          "'%s' and should be set to '%s' this has been fixed.")
```

Comment:
On second thought, I think it might be better to remove the logic from this PR which resets the permissions on the `privatekeydir` and master/agent private key pem files within it.  I had forgotten that the Puppet settings which map to this directory and files can be specified with an explicit mode in the `puppet.conf` file.  From https://docs.puppetlabs.com/references/latest/configuration.html#configuration-settings:

> Settings that take a single file or directory can optionally set the owner, group, and mode for their value: rundir = $vardir/run { owner = puppet, group = puppet, mode = 644 }

If we were to reset these permissions at startup, we could overwrite an alternate mode that the user may have set in their `puppet.conf` file.  To avoid this, I don't think we should try to figure out how to have Puppet Server extract the mode value for Ruby Puppet settings and use that to set permissions.  Since Ruby Puppet code isn't managing the CA private key mode when running under Puppet Server, I do still think it would be good to retain the logic which resets the permissions to 0640 for the CA private key mode at startup.  For the `privatekeydir` and its subordinate pems, though, I now think it might be better to just let Ruby Puppet handle resetting permissions, if needed, at startup as it already does / knows how to do.  From Puppet Server's perspective, then, the only thing we'd additionally need to handle for the `privatekeydir` and subordinate pems is setting world-inaccessible permissions as the directory/files are first generated.